[6/n][torch/elastic][upstream] Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api

aivanou opened the pull request:
Summary: Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api

Test Plan:
buck test mode/dev-nosan //pytorch/elastic/torchelastic/...
    buck test mode/dev-nosan //caffe2/test/distributed/elastic/agent/server/test/...

Differential Revision: D27352777

cbalioglu commented:
Looks good to me. Do you mind checking the mypy error though?

mrshenli commented on test/distributed/elastic/launchers/__init__.py:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env python3
```
Would I be correct if I assume, there will be followup efforts to consolidate with `https://github.com/pytorch/pytorch/blob/master/torch/distributed/launch.py`?

mrshenli commented on test/distributed/elastic/launchers/api_test.py:
```diff
@@ -0,0 +1,247 @@
+#!/usr/bin/env python3
```
do we plan to plug `api_test.py` to `test/run_test.py` in the same as other distributed tests, or do we plan to use a separate test launching solution. If it's the former and if the tests are supposed to pass today, shall we enable them in this PR. If it's the later, we might need to make sure that we can still create an issue with title format "DISABLED testname.TestClass" to disable it in CI.

mrshenli commented on test/distributed/elastic/launchers/bin/test_script.py:
```diff
@@ -0,0 +1,68 @@
+#!/usr/bin/env python3
```
nit: existing tests in the distributed folder use the following convention: 1) file with a `test_` prefix are entry points of tests and is registered in `test/run_test.py`. And we can use things like `pytest test/distributed/rpc/test_tensorpipe_agent.py -k test_func_name -vs` to run individual tests. 2) if some test implementation needs to shared across different test runs, we create a file with `_test` for implementation and import that in multiple `test_` prefix files, e.g., `rpc_test.py` is imported by both `test_tensorpipe_agent.py` and `test_process_group_agent.py`.

I am not sure if makes sense to follow the same convention here, but might be good to do so if it's not too much efforts.

aivanou committed:
[6/n][torch/elastic][upstream] Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api (#54852)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/54852

Move torchelastic/distributed/api to torch/distributed/elastic/launchers/api

Test Plan:
buck test mode/dev-nosan //pytorch/elastic/torchelastic/...
    buck test mode/dev-nosan //caffe2/test/distributed/elastic/agent/server/test/...

Reviewed By: kiukchung

Differential Revision: D27352777

fbshipit-source-id: eb29b88833ef226cc8e6b3c1747a5d067284c3dd
